fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java

```diff
@@ -190,13 +219,90 @@ public void afterHistory(FHIRPersistenceEvent event) throws FHIRPersistenceInter
+            throws FHIRPersistenceInterceptorException {
+        if (resource instanceof Provenance) {
+            if (!isAllowed(((Provenance) resource).getTarget(), event.getPersistenceImpl(), patientIdFromToken, Permission.READ, scopesFromToken)) {
+                String msg = Permission.READ + " permission to 'Provenance/" + resource.getId() +
+                        "' with context id(s): " + patientIdFromToken +
+                        " requires access to one or more of its target resources.";
+                if (log.isLoggable(Level.FINE)) {
+                    log.fine(msg);
+                }
+                throw new FHIRPersistenceInterceptorException(msg)
+                        .withIssue(FHIRUtil.buildOperationOutcomeIssue(msg, IssueType.FORBIDDEN));
+            }
+        }
+    }
+
+    /**
+     * Determine whether authorization to one or more referenced resources is granted by the end user in the form of scope strings
+     *
+     * @param references a list of resource references; this method will dereference only relative literal references
+     * @param persistence the FHIRPersistence implementation to use for dereferencing the literal references
+     * @param contextIds an identifier for the current context (e.g. patient or user) as determined by the scope strings
+     * @param requiredPermission
+     * @param approvedScopes a list of SMART scopes associated with the request
+     * @throws IllegalStateException if the baseUrl cannot be computed from the request context
+     * @throws FHIRPersistenceInterceptorException if the interaction is not permitted
+     */
+    private boolean isAllowed(List<Reference> references, FHIRPersistence persistence, List<String> contextIds, Permission requiredPermission, List<Scope> approvedScopes) {
+        boolean allow = false;
+
+        String baseUrl;
+        try {
+            baseUrl = ReferenceUtil.getServiceBaseUrl();
+        } catch (FHIRSearchException e) {
+            throw new IllegalStateException("Unexpected error while computing the service baseUrl for ");
+        }
+
+        for (Reference ref : references) {
+            ReferenceValue referenceValue = ReferenceUtil.createReferenceValueFrom(ref, baseUrl);
+            if (ReferenceValue.ReferenceType.LITERAL_RELATIVE == referenceValue.getType()) {
+                Class<? extends Resource> resourceType = ModelSupport.getResourceType(referenceValue.getTargetResourceType());
+                try {
+                    SingleResourceResult<? extends Resource> result = executeRead(persistence, referenceValue, resourceType);
+
+                    if (result.isSuccess() && checkCompartment(result.getResource(), CompartmentType.PATIENT, contextIds)) {
+                        allow = true;
+                        break;
+                    } if (log.isLoggable(Level.FINE)){
```

Comment:
Don't you only want to log this if result.isSuccess() is false?

Reply:
oops, yes.  definitely meant that to be an else if.  good catch!
```suggestion
                    } else if (log.isLoggable(Level.FINE)){
```